dask/tests/test_base.py

```diff
@@ -1694,3 +1696,85 @@ def test_emscripten_default_scheduler(params):
         ]
     )
     proc.check_returncode()
+
+
+@pytest.mark.parametrize("format_str", ["%s", "{0}", "{result}"])
+def test_print_on_compute_delayed(capsys, format_str):
+    def inc(x):
+        return x + 1
+
+    val = dask.delayed(inc)(1)
+    val = print_on_compute(f"delayed value: {format_str}", val)
+    val.compute()
+    captured = capsys.readouterr()
+    assert captured.out == "delayed value: 2\n"
+
+
+@pytest.mark.parametrize("format_str", ["%s", "{0}", "{result}"])
+def test_print_on_compute_dataframe(capsys, format_str):
+    pytest.importorskip("dask.dataframe")
+
+    df = pd.DataFrame({"x": range(4)})
+    ddf = dd.from_pandas(df, npartitions=1)
+    ddf = print_on_compute("Original: {result}", ddf)
+    res = ddf[ddf.x > 1]
+    res = print_on_compute("Filtered: {result}", res)
+    res.compute()
+
+    captured = capsys.readouterr()
+    assert captured.out == textwrap.dedent(
+        """\
+        Original:    x
+        0  0
+        1  1
+        2  2
+        3  3
+        Filtered:    x
+        2  2
+        3  3
+        """
+    )
+
+
+def test_print_on_compute_intermediate(capsys):
+    pytest.importorskip("dask.dataframe")
+
+    df = pd.DataFrame({"x": range(1000)})
+    ddf = dd.from_pandas(df, npartitions=2)
+    ddf.x = ddf.x + 1
+    lower_bound = ddf.x.min()
+    lower_bound = print_on_compute("Lower bound {result}", lower_bound)
+    Q = ddf.x.quantile(0.75)
+    Q = print_on_compute("Quantile {result}", Q)
+    res = Q + (1.5 * Q - lower_bound)
```

Comment:
Very conveniently, this line in your test works indifferently with a dd.Scalar or with a Delayed. This is not something you should rely upon.